motionplan/tpSpaceRRT.go

```diff
@@ -250,64 +305,87 @@ func (mp *tpSpaceRRTMotionPlanner) rrtBackgroundRunner(
+				seedReached = mp.attemptExtension(ctx, goalReached.node, rrt.maps.startMap, false, mp.randseed.Int31()+mp.randseed.Int31())
 				if seedReached.error != nil {
 					rrt.solutionChan <- &rrtPlanReturn{planerr: seedReached.error, maps: rrt.maps}
 					return
 				}
-				if seedReached.node == nil {
-					continue
-				}
-				goalReached = mp.attemptExtension(ctx, seedReached.node, rrt.maps.goalMap, true)
-				if goalReached.error != nil {
-					rrt.solutionChan <- &rrtPlanReturn{planerr: goalReached.error, maps: rrt.maps}
-					return
-				}
-				if goalReached.node == nil {
-					continue
+				if seedReached.node != nil {
+					reachedDelta = mp.planOpts.DistanceFunc(&ik.Segment{
+						StartPosition: seedReached.node.Pose(),
+						EndPosition:   goalReached.node.Pose(),
+					})
+					if reachedDelta > mp.planOpts.GoalThreshold {
+						goalReached = mp.attemptExtension(ctx, seedReached.node, rrt.maps.goalMap, true, mp.randseed.Int31()+mp.randseed.Int31())
+						if goalReached.error != nil {
+							rrt.solutionChan <- &rrtPlanReturn{planerr: goalReached.error, maps: rrt.maps}
+							return
+						}
+					}
+					if goalReached.node != nil {
+						reachedDelta = mp.planOpts.DistanceFunc(&ik.Segment{StartPosition: seedReached.node.Pose(), EndPosition: goalReached.node.Pose()})
+					}
 				}
-				reachedDelta = mp.planOpts.DistanceFunc(&ik.Segment{StartPosition: seedReached.node.Pose(), EndPosition: goalReached.node.Pose()})
 			}
-			if reachedDelta <= mp.algOpts.poseSolveDist {
+			if reachedDelta <= mp.planOpts.GoalThreshold {
 				// If we've reached the goal, extract the path from the RRT trees and return
 				path := extractPath(rrt.maps.startMap, rrt.maps.goalMap, &nodePair{a: seedReached.node, b: goalReached.node}, false)
-				correctedPath, err := rectifyTPspacePath(path, mp.frame, spatialmath.NewZeroPose())
-				if err != nil {
-					rrt.solutionChan <- &rrtPlanReturn{planerr: err, maps: rrt.maps}
-					return
-				}
-				rrt.solutionChan <- &rrtPlanReturn{steps: correctedPath, maps: rrt.maps}
+				returnFinishedPath(path)
 				return
 			}
 		}
 		if iter%mp.algOpts.attemptSolveEvery == 0 {
-			// Attempt a solve; we exhaustively iterate through our goal tree and attempt to find any connection to the seed tree
+			// Attempt a solve; we iterate through our goal tree and attempt to find any connection to the seed tree
 			paths := [][]node{}
-			for goalMapNode := range rrt.maps.goalMap {
-				seedReached := mp.attemptExtension(ctx, goalMapNode, rrt.maps.startMap, false)
+
+			// Exhaustively searching the tree gets expensive quickly, so we cap the number of connect attempts we make each time we call
+			// this.
+			attempts := 0                                                               // Track the number of connection attempts we have made
+			pctCheck := float64(defaultMaxConnectAttempts) / float64(len(mp.goalNodes)) // Target checking this proportion of nodes.
+
+			for _, goalMapNode := range mp.goalNodes {
```

Comment:
The neighbor manager is a one-to-many comparison.

Doing an exhaustive many-to-many is what's causing timeouts and perf issues, so we don't want that.

Reply:
We're trying a many to many right now and it seems like we could have performance improvements if instead of actually trying those extensions we use the neighnor manager to inform what would be some of the likely candidates.

Also, in general I'm a little confused why we're even doing this in the first place. In theory we should be able to connect the two trees just by the natural extensions we're doing in the first place

Reply:
> In theory we should be able to connect the two trees just by the natural extensions we're doing in the first place

In theory we eventually would, yes, but in practice this leads to us solving faster. When we extend the RRT trees, both are extending towards some random point, and in constrained areas like a slam map this can result in trees extending past some window where they could connect. Such a connection would never be made during normal RRT operation unless we luck into juuuust the right random node.

I'm not sure how the neighbor manager would be useful.

Reply:
Note that for tpspace, the neighbor manager needs to operate in terms of PTG trajectories, not simple cartesian space. So running the neighbor manager actually runs IK across the query tree.

This is why RSDK-6052 exists.

Something that could be a neat speed and quality improvement would be to create multiple paths here, where far apart nodes just use cartesian nearness, while close-together ones check for ptg viability, but that's definitely beyond the scope of this ticket.

Reply:
More to the point, the reason that this change needs to go into this PR, is that with the long-distance solves that are now possible, the trees can get larger quicker, due to the intermediate nodes we add along the way; we need to limit the amount of time we spend doing these connection attempts more now.